### PR TITLE
[WD-13151] chore: split last seen column into "Status" and "Last Heartbeat" colu…

### PR DESCRIPTION
**Pull request by Kxiru**
…mns.

Work done:
- Split columns into "Status" and "Last Heartbeat"
- Altered Heartbeat logic to display last seen times differently.
- Created a new ClusterStatus Component.
- Moved Cluster metrics into dedicated cluster_metrics folder.
- Changed Site Generation last-seen attribute to generate random times within the last 30 minutes for visualisation purposes.

image

### CODE REVIEW

File: ui/src/pages/clusters/ClusterList.tsx [hunk 1]
```diff
@@ -6,12 +6,13 @@ import { MainTable, Row, TablePagination } from "@canonical/react-components";
 import Loader from "components/Loader";
 import BaseLayout from "components/BaseLayout";
 import { Link } from "react-router-dom";
-import { ClusterInstances } from "components/meters/ClusterInstances";
-import { ClusterCpu } from "components/meters/ClusterCpu";
-import { ClusterMemory } from "components/meters/ClusterMemory";
-import { ClusterDisk } from "components/meters/ClusterDisk";
-import { ClusterNodes } from "components/meters/ClusterNodes";
-import ClusterHeartbeat from "components/ClusterHeartbeat";
+import { ClusterInstances } from "components/cluster_metrics/ClusterInstances";
```
Comment (edlerd) [inline ui/src/pages/clusters/ClusterList.tsx]:
```suggestion
import { ClusterInstances } from "./metrics/ClusterInstances";
```
Let's keep those under the pages/clusters/metrics folder. That way they have a shorter and more concise path and are easier to find. Components is meant for generic ones, which are not tied to any dedicated page / functionality.

File: ui/src/util/helpers.tsx [hunk 2]
```diff
@@ -17,17 +17,26 @@ export const handleResponse = async (response: Response) => {
...
+    returnArray[0] =
+      lastHeartbeatHrs < 1 //Displayed for "Last Seen"'s of 5-59 Minutes
+        ? `< 1 hour ago`
+        : lastHeartbeatHrs < 2
+          ? `1 hour ago`
+          : `${lastHeartbeatHrs} hours ago`;
   }
 
-  return "Unresponsive";
+  return returnArray;
```
Comment (edlerd) [inline ui/src/util/helpers.tsx]: I think this function is doing too many things. Maybe we refactor it to return the `lastHeartbeatMins`. The logic to figure out the copy for `ClusterHeartbeat` can be local to that component. And the logic to calculate the status / icon can be local to the `ClusterStatus` component.

**Review by edlerd — CHANGES_REQUESTED** [review]
QA lgtm, two small ideas to simplify the structure.

File: ui/src/util/helpers.tsx (hunk 2 above)
Comment (Kxiru) [reply ui/src/util/helpers.tsx]: Thanks, for the review, David. All done!

File: ui/src/pages/clusters/metrics/ClusterStatus.tsx [hunk 3]
```diff
@@ -7,20 +7,21 @@ interface Props {
...
 
   return (
     <div>
       <Icon
         className="is-light p-side-navigation__icon"
         name={`status-${heartbeatClass}`}
       />
-      {heartbeatStatus}
+      {/* Status displays as Online when the cluster was last seen less than 5 minutes ago. */}
+      {lastHeartbeatMins ? "Online" : "Offline"}
```
Comment (edlerd) [inline ui/src/pages/clusters/metrics/ClusterStatus.tsx]:
```suggestion
      {lastHeartbeatMins < 5 ? "Online" : "Offline"}
```

File: ui/src/util/helpers.tsx [hunk 4]
```diff
@@ -17,17 +17,11 @@ export const handleResponse = async (response: Response) => {
 export const isWidthBelow = (width: number): boolean =>
   window.innerWidth < width;
 
-export function getHeartbeatStatus(cluster: Cluster): string {
+export function getHeartbeatStatus(cluster: Cluster): number {
```
Comment (edlerd) [inline ui/src/util/helpers.tsx]:
```suggestion
export function getMinuesSinceLastHeartbeat(cluster: Cluster): number {
```

File: ui/src/util/helpers.tsx [hunk 5]
```diff
@@ -17,17 +17,11 @@ export const handleResponse = async (response: Response) => {
...
   const lastSeenTime = Date.parse(cluster.last_status_update_at);
-  const lastHeartbeat = Math.floor((Date.now() - lastSeenTime) / 60000); //Value in Minutes
+  const lastHeartbeatMins = Math.floor((Date.now() - lastSeenTime) / 60000);
```
Comment (edlerd) [inline ui/src/util/helpers.tsx]:
```suggestion
  return Math.floor((Date.now() - lastSeenTime) / 60000);
```

File: ui/src/pages/clusters/metrics/ClusterStatus.tsx (hunk 3 above)
Comment (edlerd) [reply ui/src/pages/clusters/metrics/ClusterStatus.tsx]: I think this is missing a comparison.

File: ui/src/util/helpers.tsx (hunk 4 above)
Comment (edlerd) [reply ui/src/util/helpers.tsx]: More precise?

File: ui/src/util/helpers.tsx (hunk 5 above)
Comment (edlerd) [reply ui/src/util/helpers.tsx]: Can simplify a bit.

**Review by edlerd — DISMISSED** [review]
LGTM :+1:

**Review by edlerd — APPROVED** [review]

**Pull request merged**